ERCS/erc-7786.md

```diff
@@ -0,0 +1,217 @@
+
+The gateway MUST emit a `MessageSent` event with the appropriate identifier once no more post-processing is required by the source gateway. Further post-processing MAY be required in the source or destination gateways.
+
+### Destination Gateway
+
+A Destination Gateway is a contract that implements a protocol to validate messages sent on other chains and have them received at their destination.
+
+The gateway can operate in Active or Passive Mode.
+
+In both modes, the destination account of a message, aka the receiver, MUST implement a `receiveMessage` function. The use of this function depends on the mode of the gateway as described in the following sections.
+
+```solidity
+interface IERC7786Receiver {
+    function receiveMessage(
+        address gateway,
+        bytes calldata gatewayMessageKey,
+        string calldata source, // CAIP-2 chain identifier
+        string calldata sender, // CAIP-10 account address
+        bytes calldata payload,
+        bytes[] calldata attributes
+    ) external payable;
+}
+```
+
+#### Active Mode
+
+The gateway directly invokes `receiveMessage`, and only does so with valid messages. The receiver MUST assume that a message is valid if the caller is a known gateway.
+
+The arguments `gateway` and `gatewayMessageKey` are unused in active mode and SHOULD be zero and empty respectively.
+
+#### Passive Mode
+
+The gateway does not directly invoke `receiveMessage`, but provides a means to validate messages. The receiver allows any party to invoke `receiveMessage`, but if the caller is not a known gateway it MUST check that the gateway provided as an argument is a known gateway, and it MUST validate the message against it before accepting it, forwarding the message key.
+
+A gateway acting in passive mode MUST implement `IERC7786GatewayDestinationPassive`. If a gateway operates exclusively in active mode, the implementation of this interface is OPTIONAL.
+
+```solidity
+interface IERC7786GatewayDestinationPassive {
+    function validateReceivedMessage(
+        bytes calldata messageKey,
+        string calldata source, // CAIP-2 chain identifier
+        string calldata sender, // CAIP-10 account address
+        bytes calldata payload,
+        bytes[] calldata attributes
+    ) external;
+}
+```
+
+##### `validateReceivedMessage`
+
+Checks that there is a valid and as yet unexecuted message whose contents are exactly those passed as arguments and whose receiver is the caller of the function. The message key MAY be an identifier, or another piece of data necessary for validation.
+
+MUST revert if the message is invalid or has already been executed.
+
+TBD: Passing full payload or payload hash (as done by Axelar). Same question for attributes, possibly different answer depending on attribute malleability.
+
```

Comment:
I think this is a good place for the properties section. We could call it guarantees.

- Only once execution
- Safety: A message is delivered at the destination if and only if it was sent at the source, and the sending transaction is finalized.
- Liveness: A sent message is delivered at the destination eventually, with minimal delay.
- Decentralization: The above properties do not assume trust of any centralized entity. Safety should be ensured through a trustless mechanism such as proofs or a decentralized, open validator set. Liveness should be ensured through permissionless relaying (anyone can relay), and a decentralized, open validator set (if applicable).

Can expand a little bit on this, and maybe in another doc go into greater detail. But I think this is important to specify here, because we don't want totally centralized entities implementing this API.

Reply:
Added more or less this below.